### ISSUE TITLE
Bag.from_s3

### ISSUE THREAD

**Issue by mrocklin** [ContinuumIO/dask#241]
My understanding is that people often compute on EC2 on data by pulling it directly from S3, skipping local HDFS altogether.  This is great for us because it avoids our shortcomings in data locality.  We should probably use a tool like `boto` to pull text data from S3 into a Bag.  The `dask.bag.from_hdfs` function and and `odo.aws` module can serve as useful models here.

This would presumably be useful in the construction of a dask.distributed demo.

**Comment by mrocklin** [comment]
The goal here is to start with an S3 bucket (or key) and end up with a bag of lines of text

```python
>>> b = db.from_s3('s3://bucket...')
>>> b.take(3)
('line of text',
 'line of text',
 'line of text')
```

Historically I have integrated decompression into this function (e.g. if the name ends with `.gz` then run the downloaded data through `stream_decompress`) haven't tried to integrate parsing (e.g. still return text even if the filename ends with `.json`).  This differs a bit from odo interaction but I think that this is more appropriate for dask.  The user can compose with their parsing function of choice with a `.map(json.loads)` afterwards if they like.

**Comment by shoyer** [comment]
Would these methods return bytes or unicode on Python 3? I presume the former.

This could actually be quite useful for us -- we use S3 for storing all of our bulk data. The complementary to_s3 method would also be handy. 

Have you thought about the story for converting between different dask containers? Ultimately, I'd love to be able to do S3 -> dask.bag of raw source files -> xray.Dataset of dask.array objects -> dask.bag of netcdf bytes -> S3. This sort of thing is a fairly common workflow for us.

**Comment by mrocklin** [comment]
For dask.bag I've only interacted with text data on HDFS/S3 so I would tend to lean more towards unicode.  However, if pulling non-text-based data off of S3 is common then we'd clearly need to expand out the API.  Perhaps `from_textfiles` should dispatch based on protocol like `odo` does.  

Regarding storing to S3 can you provide feedback on the existing [`to_textfiles`](http://dask.readthedocs.org/en/latest/bag.html#dask.bag.core.Bag.to_textfiles) method, in particular the API to specify filenames in text data (docstring very recently updated.)

Regarding converting between dask containers yes, I think that this is an important and relatively untapped improvement.  There is a `Bag.to_dataframe` method currently and Array<->DataFrame shouldn't be too hard.  I'm not sure what the standard mechanism for Bag -> Array should be; you presumably have more experience there. 

It's also possible that bag isn't the right abstraction for what you want to do.  I suspect that you probably have a bunch of binary blobs and you want to think of each one as being in its own "chunk."  This is a bit different from the bag case where a chunk is a list of presumably multiple elements.  One can use bags in this case (lists of single elements) but it would also be pretty easy to make an even simpler collection type.

At the moment I'm only thinking about S3 loading based on my experience, which is limited to line-delimited JSON.  If we can add a few different use cases / perspectives then we might come up with a better API.  @shoyer can I ask you to mock out the sort of functionality you'd like to see in a fake example?

Finally, there is always the "just make a dict" release valve for custom load/store.  One can always make a dask `Array` with a dask graph, `chunks`, and `dtype`.

**Comment by shoyer** [comment]
For our use cases, we are equally likely to be pulling a CSV or a netCDF or a geotiff or some other opaque binary format (there are a lot of these). Something like `open`'s `mode` argument would be useful here.

The `to_textfiles` API looks prtty solid to me.

> I suspect that you probably have a bunch of binary blobs and you want to think of each one as being in its own "chunk." 

This is correct.

Let me check with some of my co-workers and see if we can mock something out. It is indeed quite plausible that we may just want to write task graphs directly.

**Comment by mrocklin** [comment]
An example using Boto 

```python
In [1]: import boto

In [2]: conn = boto.connect_s3()

In [3]: bucket = conn.get_bucket('nyqpug')

In [4]: key = bucket.get_key('tips.csv')

In [5]: key.read(100)
Out[5]: 'total_bill,tip,sex,smoker,day,time,size\n16.99,1.01,Female,No,Sun,Dinner,2\n10.34,1.66,Male,No,Sun,Din'

In [6]: key.read(100)
Out[6]: 'ner,3\n21.01,3.5,Male,No,Sun,Dinner,3\n23.68,3.31,Male,No,Sun,Dinner,2\n24.59,3.61,Female,No,Sun,Dinner'

In [7]: key.read(100)
Out[7]: ',4\n25.29,4.71,Male,No,Sun,Dinner,4\n8.77,2.0,Male,No,Sun,Dinner,2\n26.88,3.12,Male,No,Sun,Dinner,4\n15.'
```

**Comment by mrocklin** [comment]
Closed by #245
